File: scala/src/P02Recursion.scala
object P02Recursion {
  def main(args: Array[String]): Unit = {
    val xs = List(5, 3, 90, -1, 2)

    assert(max(xs) == 90)
    assert(max(List()) == 0)
  }

  def max(xs: List[Int]): Int = {
    xs match {
      case List() => 0
      case List(a) => a
      case a :: as =>
        val maxTail = max(as)
        if (a > maxTail) a else maxTail
    }
  }
}

File: java/src/P02Recursion.java
import java.util.List;

public class P02Recursion {
    public static void main(String[] args) {
        var xs = List.of(5, 3, 90, -1, 2);

        assert 90 == max(xs);
        assert 0 == max(List.of());
    }

    private static Integer max(List<Integer> xs) {
        if (xs.isEmpty()) return 0;
        else if (xs.size() == 1) return xs.get(0);
        else {
            var head = xs.get(0);
            var tail = xs.subList(1, xs.size());
            var maxTail = max(tail);
            if (head > maxTail) return head;
            else return maxTail;
        }
    }
}